### ISSUE TITLE
如何利用GPU进行加速训练？

### ISSUE THREAD

**Issue by zhouaihui** [secretflow/secretflow#180]
### Issue Type

Feature Request

### Source

binary

### Secretflow Version

0.7.7b1

### OS Platform and Distribution

mac 

### Python version

3.8

### What happend and What you expected to happen.

```shell
如何利用GPU进行加速训练？
```

### Reproduction code to reproduce the issue.

```shell
如何利用GPU进行加速训练？
```

**Comment by zhouaihui** [comment]
hi, @cs1317，关于GPU加速联邦学习

以TensorFlow为例
按照[官方解释](https://www.tensorflow.org/guide/gpu?hl=zh-cn)，对于同时具有`CPU:0` 和 `GPU:0`的设备，并且计算kernel同时支持CPU和GPU，则会默认使用GPU。
所以结论是SecretFlow内置的TensorFlow版本联邦学习算法可以享受TF的上述特性。

如果你需要使用特定的GPU来进行计算，则目前SecretFlow内置的联邦算法还没有暴露GPU相关参数。不过好消息是你可以手动编写你自己的算法来完成这个目的。
以官方示例代码为例，下列代码会在alice所在机器的指定GPU上执行
```python
alice = sf.PYU('alice')

def gpu_run():
  tf.debugging.set_log_device_placement(True)
  
  # Place tensors on the CPU
  with tf.device('/CPU:0'):
    a = tf.constant([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
    b = tf.constant([[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]])
  
  c = tf.matmul(a, b)
  print(c)

alice(gpu_run)()
```

**Comment by cs1317** [comment]
TensorFlow和PyTorch在下载的时候都会区分GPU版和CPU版的，secretflow内置的TensorFlow和PyTorch是否会判断机器的GPU是否可用，从而自动下载GPU/CPU版的TensorFlow/PyTorch？
还是说，我可以自己先下载好GPU版的TensorFlow/PyTorch，然后再下载secretflow?

**Comment by zhouaihui** [comment]
如果你是通过`pip install secretflow`安装或者使用我们提供的docker，则默认安装的是支持GPU的TF/Torch。

**Comment by cs1317** [comment]
通过pip install secretflow 下载的tensorflow，GPU不可用；pytorch的GPU可用
`>>> import tensorflow as tf
>>> print(**tf.test.is_gpu_available()**)
WARNING:tensorflow:From <stdin>:1: is_gpu_available (from tensorflow.python.framework.test_util) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.config.list_physical_devices('GPU')` instead.
2022-11-04 10:33:11.198825: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcusolver.so.11'; dlerror: libcusolver.so.11: cannot open shared object file: No such file or directory
2022-11-04 10:33:11.199726: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1850] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
**False**
>>> import torch
>>> print(**torch.cuda.is_available()**)
**True**`

**Comment by zhouaihui** [comment]
请参考[GPU支持](https://www.tensorflow.org/install/gpu?hl=zh-cn)，目前SecretFlow依赖的tensorflow2.10版本，TF2是同时支持CPU和GPU，当你的机器支持GPU且正确安装了cuda的话，应该会默认选择GPU执行。

ps： tensorflow-gpu是仅支持GPU的版本。

**Comment by cs1317** [comment]
我使用pytorch进行训练，一直卡在这:
100%|██████████| 48/48 [00:01<00:00, 39.24it/s]#  
很长时间无法进入训练，nvidia-smi显示GPU无占用
我想知道是什么原因导致的
`history = fl_model.fit(train_data,train_label,validation_data=(test_data, test_label),epochs=20,batch_size=32,
aggregate_freq=1)`

**Comment by zhouaihui** [comment]
hi，PyTorch使用GPU需要显示指定，抱歉目前封装好的FLModel里的torch代码默认是CPU写法。
如果需要使用pytorch运行在gpu上，目前可以手动开发编写训练代码，具体可以参考[PyTorch Cuda](https://pytorch.org/docs/stable/cuda.html)。

**Comment by zhouaihui** [comment]
可以把完整代码发一下，我们尝试复现下。

**Comment by cs1317** [comment]
[ImageClassfyPytorch_copy.py.zip](https://github.com/secretflow/secretflow/files/9957592/ImageClassfyPytorch_copy.py.zip)
图像分类任务，数据集是一个flower数据集，5个类,
https://link.zhihu.com/target=https%3A//storage.googleapis.com/download.tensorflow.org/example_images/flower_photos.tgz
我将图像和标签转成numpy, 拆分后分别保存在两台机器上

**Comment by zhouaihui** [comment]
麻烦贴一下代码

**Comment by cs1317** [comment]
`import secretflow as sf
import spu
from secretflow.data.ndarray import load
from secretflow.security import SPUAggregator

#ray start --head --node-ip-address="ip1" --port="6379" --resources='{"alice": 8}' --include-dashboard=False --disable-usage-stats
#ray start --address="ip1:6379" --resources='{"bob": 8}' --disable-usage-stats

sf.shutdown()

sf.init(address="ip1:6379")

cluster_def = {
        'nodes': [
            {
                'party': 'alice',
                'id': '0',
                'address': 'ip1:8285',
            },
            {
                'party': 'bob',
                'id': '1',
                'address': 'ip2:8285',
            },
        ],
        'runtime_config': {
            'protocol': spu.spu_pb2.SEMI2K,
            'field': spu.spu_pb2.FM128,
            'sigmoid_mode': spu.spu_pb2.RuntimeConfig.SIGMOID_REAL,
        }
    }

spu = sf.SPU(cluster_def=cluster_def)
alice=sf.PYU("alice")
bob=sf.PYU("bob")

train_data = load({bob: '/fds/chengsai1/x_train1.npy', alice: '/chengsai1/chengsai1/privacy_detection/secretflow-test/x_train2.npy'})
test_data  = load({bob: '/fds/chengsai1/x_test1.npy',  alice: '/chengsai1/chengsai1/privacy_detection/secretflow-test/x_test2.npy'} )
train_label = load({bob: '/fds/chengsai1/y_train1.npy', alice: '/chengsai1/chengsai1/privacy_detection/secretflow-test/y_train2.npy'})
test_label  = load({bob: '/fds/chengsai1/y_test1.npy',  alice: '/chengsai1/chengsai1/privacy_detection/secretflow-test/y_test2.npy'} )
print("data load successfully!")



#使用指定的GPU
import torch
torch.cuda.set_device(1)

from torch import nn, optim
from secretflow.ml.nn.fl.backend.torch.utils import BaseModule, TorchModel
from secretflow.ml.nn.fl.utils import metric_wrapper, optim_wrapper
from secretflow.security.aggregation import SecureAggregator
from secretflow.ml.nn import FLModel
from torchmetrics import Accuracy, Precision

class ConvNet(BaseModule):
    def __int__(self,num_classes):
        super(ConvNet,self).__init__()
        self.model=nn.Sequential(
            nn.Conv2d(3,32,kernel_size=(3,3)),
            nn.ReLU(),
            nn.BatchNorm2d(),
            nn.MaxPool2d((2,2)),
            nn.Conv2d(32, 64, kernel_size=(3, 3)),
            nn.ReLU(),
            nn.BatchNorm2d(),
            nn.MaxPool2d((2, 2)),
            nn.Flatten(),
            nn.Dropout(),
            nn.Linear(num_classes),
            nn.Softmax()
        )

    def forward(self,x):
        print(x.shape)
        x=self.model(x)
        return x

loss_fn = nn.CrossEntropyLoss
optim_fn = optim_wrapper(optim.Adam, lr=1e-2)
model_def = TorchModel(
    model_fn=ConvNet,
    loss_fn=loss_fn,
    optim_fn=optim_fn,
    metrics=[
        metric_wrapper(Accuracy, num_classes=10, average='micro'),
        metric_wrapper(Precision, num_classes=10, average='micro'),
    ],
)
print("model create")

device_list = [alice, bob]
server = alice
aggregator = SecureAggregator(server,[alice,bob])
print("聚合success")

# spcify params
fl_model = FLModel(
    server=server,
    device_list=device_list,
    model=model_def,
    aggregator=aggregator,
    strategy='fed_avg_w', # fl strategy
    backend="torch", # backend support ['tensorflow', 'torch']
)
print("FLModel load success")

history = fl_model.fit(
            train_data,
            train_label,
            validation_data=(test_data, test_label),
            epochs=20,
            batch_size=32,
            aggregate_freq=1,
        )
print("train finished")

fl_model.save_model({alice:"/chengsai1/chengsai1/privacy_detection/secretflow-test/flower_classification.pth"})

import matplotlib
matplotlib.use('AGG')#或者PDF, SVG或PS
from matplotlib import pyplot as plt


# Draw accuracy values for training & validation
plt.plot(history.global_history['accuracy'])
plt.plot(history.global_history['val_accuracy'])
plt.title('FLModel accuracy')
plt.ylabel('Accuracy')
plt.xlabel('Epoch')
plt.legend(['Train', 'Valid'], loc='upper left')
plt.savefig("accuracy.png")

plt.plot(history.global_history['loss'])
plt.plot(history.global_history['val_loss'])
plt.title('FLModel loss')
plt.ylabel('Loss')
plt.xlabel('Epoch')
plt.legend(['Train', 'Valid'], loc='upper left')
plt.savefig('loss.png')`
如果看着比较乱，可以下载代码
https://github.com/secretflow/secretflow/files/9957592/ImageClassfyPytorch_copy.py.zip

**Comment by zhouaihui** [comment]
收到，我们先尝试复现下。

**Comment by zhangxingmeng** [comment]
btw，cpu版本会卡主吗?

**Comment by cs1317** [comment]
会的

**Comment by cs1317** [comment]
[flower数据集](https://storage.googleapis.com/download.tensorflow.org/example_images/flower_photos.tgz)
已改正，但仍然会卡在这
![train](https://user-images.githubusercontent.com/50541363/201616437-875835ae-cd56-432e-a015-38536af3d60d.png)

**Comment by zhangxingmeng** [comment]
你好,我有两个建议:
1、在ray.init的时候把num_cpus参数调大一点试试（hang住大概率这样是可以解决的）
2、先用单机把这个模型跑通，然后在迁移到secretflow上，在定义模型的时候吧num_class给固定住试试
![image](https://user-images.githubusercontent.com/103552200/201659550-be4204ad-bb4d-4bd5-8f97-0cd2eb5f51fc.png)
因为我这边看你的num_clasnum_class并没有设置，所以怀疑可能这里也会有问题。

**Comment by ceasona** [comment]
默认是不支持GPU的，ray remote的时候需要添加gpu相关参数，其次，model和data都需要加载到gpu上才行

**Comment by zhouaihui** [comment]
支持GPU和ray没有强依赖关系。隐语本身并不直接暴露ray给用户使用。

> 默认是不支持GPU的，ray remote的时候需要添加gpu相关参数，其次，model和data都需要加载到gpu上才行

**Comment by ceasona** [comment]
> 支持GPU和ray没有强依赖关系。隐语本身并不直接暴露ray给用户使用。
> 
> > 默认是不支持GPU的，ray remote的时候需要添加gpu相关参数，其次，model和data都需要加载到gpu上才行

不对吧，模型训练依赖于ray，ray控制着资源

**Comment by zhouaihui** [comment]
有两种形态使用GPU。一种是直接使用ray调度GPU计算任务，由ray负责GPU的资源分配。另一种是仅把ray作为计算任务的拉起框架，计算任务是以进程方式运行，计算任务可以自行决定是否访问GPU。
目前SecretFlow的FL等算法其实是第二种，ray只是负责调度FL算法实体作为ray的actor启动，算法实体可以自行负责去访问GPU。

**Comment by ceasona** [comment]
> 有两种形态使用GPU。一种是直接使用ray调度GPU计算任务，由ray负责GPU的资源分配。另一种是仅把ray作为计算任务的拉起框架，计算任务是以进程方式运行，计算任务可以自行决定是否访问GPU。 目前SecretFlow的FL等算法其实是第二种，ray只是负责调度FL算法实体作为ray的actor启动，算法实体可以自行负责去访问GPU。

哦哦，感谢解答。我在backend用torch的情况下，不改ray配置的话好像没法使用到GPU

**Comment by zhangzp9970** [comment]
我也遇到了相同的问题，在隐语中使用pytorch无法使用GPU计算。如果按照官网上的代码，只能使用CPU，非常的慢。如果在模型定义class中设置.cuda()则会直接报错说找不到GPU。但是运行torch.cuda.is_available()是true

**Comment by zhouaihui** [comment]
pytorch需要设置ray，可以尝试
1. 如果是本地模拟(address=‘local’)，则可以在sf.init时设置num_gpus，参考[sf.init](https://www.secretflow.org.cn/docs/secretflow/latest/en-US/source/secretflow#secretflow.init)
2. 如果是手动先启动ray集群，则在启动ray的命令行中添加`--num-gpus 2`（数值可自定义）

**Comment by zhangzp9970** [comment]
> pytorch需要设置ray，可以尝试
> 
> 1. 如果是本地模拟(address=‘local’)，则可以在sf.init时设置num_gpus，参考[sf.init](https://www.secretflow.org.cn/docs/secretflow/latest/en-US/source/secretflow#secretflow.init)
> 2. 如果是手动先启动ray集群，则在启动ray的命令行中添加`--num-gpus 2`（数值可自定义）

我刚试了一下，好像并不太行。这是我的运行结果，显然它还在cpu上运行
<img width="1920" alt="image" src="https://github.com/secretflow/secretflow/assets/56378403/a80bcfca-25f3-4a1e-9e25-3e6ef263ed6b">

**Comment by iven-j** [comment]
> > pytorch需要设置ray，可以尝试
> > 
> > 1. 如果是本地模拟(address=‘local’)，则可以在sf.init时设置num_gpus，参考[sf.init](https://www.secretflow.org.cn/docs/secretflow/latest/en-US/source/secretflow#secretflow.init)
> > 2. 如果是手动先启动ray集群，则在启动ray的命令行中添加`--num-gpus 2`（数值可自定义）
> 
> 我刚试了一下，好像并不太行。这是我的运行结果，显然它还在cpu上运行 <img alt="image" width="1920" src="https://user-images.githubusercontent.com/56378403/270285471-a80bcfca-25f3-4a1e-9e25-3e6ef263ed6b.png">

您好请问您解决了这个问题吗？我刚刚试着在sf.init里加入num_gpu，但是还是没有使用gpu
